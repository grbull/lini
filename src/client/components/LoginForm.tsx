import cn from 'classnames';
import React, { ChangeEvent, ReactElement, useEffect, useState } from 'react';

import { isValidEmail } from '../utils/isValidEmail';
import { ErrorMessage } from './ErrorMessage';

interface Props {
  onSubmit: (email: string) => void;
}

export function LoginForm({ onSubmit }: Props): ReactElement {
  const [email, setEmail] = useState('');
  const [isEmailValid, setIsEmailValid] = useState<boolean | undefined>(
    undefined
  );

  useEffect(() => {
    const timer = setTimeout(() => {
      if (email.length === 0) {
        setIsEmailValid(undefined);
      } else if (isValidEmail(email)) {
        setIsEmailValid(true);
      } else {
        setIsEmailValid(false);
      }
    }, 500);

    return (): void => clearTimeout(timer);
  }, [email]);

  function changeHandler(event: ChangeEvent<HTMLInputElement>): void {
    setEmail(event.target.value);
  }

  function clickHandler(): void {
    onSubmit(email);
  }

  return (
    <form>
      <p className={cn('mb-6')}>
        Enter your email address and you will be sent a link to login with.
      </p>
      {isEmailValid === false && (
        <ErrorMessage className={cn('mb-3')}>
          The email address is invalid.
        </ErrorMessage>
      )}
      <input
        aria-label="email-input"
        autoComplete="email"
        className={cn(
          'w-full',
          'mb-6',
          'p-2',
          'text-gray-600',
          'bg-gray-200',
          'border-2',
          'border-solid',
          {
            'border-gray-200': isEmailValid === undefined,
            'border-green-400 dark:border-green-200 ': isEmailValid === true,
            'border-red-200': isEmailValid === false,
          },
          {
            'hover:border-blue-300': isEmailValid === undefined,
            'hover:border-green-300': isEmailValid === true,
            'hover:border-red-300': isEmailValid === false,
          },
          {
            'focus:border-blue-300': isEmailValid === undefined,
            'focus:border-green-300': isEmailValid === true,
            'focus:border-red-300': isEmailValid === false,
          }
        )}
        name="email"
        onChange={changeHandler}
        placeholder="Email"
        value={email}
      />
      <div className={cn('flex', 'items-center')}>
        <button
          className={cn(
            'w-28',
            'py-2',
            'mx-auto',
            'border-2',
            'border-solid',
            'border-green-400 dark:border-green-200',
            'text-green-400 dark:text-green-200',
            'hover:border-green-500 dark:hover:border-green-300',
            'hover:text-green-500 dark:hover:text-green-300',
            'disabled:opacity-50',
            'dark:disabled:border-gray-200',
            'dark:disabled:text-gray-200',
            'disabled:cursor-not-allowed'
          )}
          disabled={!isEmailValid}
          onClick={clickHandler}
          type="button"
        >
          Request Link
        </button>
      </div>
    </form>
  );
}
