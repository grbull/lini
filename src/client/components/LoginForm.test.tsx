/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom/extend-expect';

import {
  fireEvent,
  render,
  RenderResult,
  waitFor,
} from '@testing-library/react';
import React from 'react';

import { LoginForm } from './LoginForm';

function setup(): RenderResult & {
  onSubmit: typeof jest.fn;
  input: HTMLInputElement;
  button: HTMLButtonElement;
} {
  const onSubmit = jest.fn();
  const utils = render(<LoginForm onSubmit={onSubmit} />);
  const input = utils.getByLabelText('email-input') as HTMLInputElement;
  const button = utils.getByText('Request Link') as HTMLButtonElement;
  return {
    onSubmit,
    input,
    button,
    ...utils,
  };
}

describe('LoginForm Component', () => {
  it('matches the snapshot', () => {
    const { asFragment } = setup();

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles input', () => {
    const { input } = setup();

    fireEvent.change(input, { target: { value: 'test@email.com' } });

    expect(input.value).toBe('test@email.com');
  });

  it('handles correct values', async () => {
    const { onSubmit, button, input } = setup();

    expect(button).toBeDisabled();

    fireEvent.change(input, { target: { value: 'test@email.com' } });

    await waitFor(() => expect(button).toBeEnabled());

    fireEvent.click(button);

    expect(onSubmit).toHaveBeenCalled();
  });

  it('handles incorrect values', async () => {
    const { onSubmit, button, input } = setup();

    expect(button).toBeDisabled();

    fireEvent.change(input, { target: { value: 'testemail.com' } });
    await waitFor(() => expect(input.value).toBe('testemail.com'));

    expect(button).toBeDisabled();

    fireEvent.click(button);

    expect(onSubmit).toHaveBeenCalledTimes(0);
  });
});
