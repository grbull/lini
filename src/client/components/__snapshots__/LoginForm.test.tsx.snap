// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component matches the snapshot 1`] = `
<DocumentFragment>
  <form>
    <p
      class="mb-6"
    >
      Enter your email address and you will be sent a link to login with.
    </p>
    <input
      aria-label="email-input"
      autocomplete="email"
      class="w-full mb-6 p-2 text-gray-600 bg-gray-200 border-2 border-solid border-gray-200 hover:border-blue-300 focus:border-blue-300"
      name="email"
      placeholder="Email"
      value=""
    />
    <div
      class="flex items-center"
    >
      <button
        class="w-28 py-2 mx-auto border-2 border-solid border-green-400 dark:border-green-200 text-green-400 dark:text-green-200 hover:border-green-500 dark:hover:border-green-300 hover:text-green-500 dark:hover:text-green-300 disabled:opacity-50 dark:disabled:border-gray-200 dark:disabled:text-gray-200 disabled:cursor-not-allowed"
        disabled=""
        type="button"
      >
        Request Link
      </button>
    </div>
  </form>
</DocumentFragment>
`;
